### ISSUE TITLE
Alternative to services based class loader discovery

### ISSUE THREAD

**Issue by JoeHegarty** [lukehutch/fast-classpath-scanner#56]
Hi,

Great project, thanks for the hard work!

@johnou and I were trying to integrate with the Orbit project (https://github.com/orbit/orbit) to replace our usage of Guava.

Unfortunately, the services based discovery of the class loaders in ClasspathFinder doesn't play nicely when the project is shaded as the package names are then incorrect. This is particularly a problem during a maven build.

It would be really nice if there was a way to specify the loaders manually for situations where the services based discovery doesn't work. If we could do this when creating the FastClasspathScanner the shading would take effect correctly and everything would work.

I'm happy to look into making the change myself, but wanted to run it by you before I started.

Thanks,
Joe

**Comment by lukehutch** [comment]
Hi, what do you mean by "when the project is shaded"?

FastClasspathScanner does make the assumption that classes must appear on a relative path (relative to a classpath directory or jar) that corresponds with the class' package name, as required by Java conventions. Is this somehow violated in your case? Or are the packages of classes changed in your case?

I'm happy to add a different classloader mechanism (ultimately that is one of the goals -- to support all common classpath / classloader variants out there). But hopefully there's a way to get ServiceLoaders to work. (e.g. if you have a packaging step that changes the packages of classes, it should also change the package names inside the service name file.)

**Comment by JoeHegarty** [comment]
Hi, sure, I can give some more context around exactly what problem we're facing.

We're hitting the exception in ClasspathFinder:
```
        if (classLoaderHandlers.isEmpty()) {
            // Should not happen, since FastClasspathScanner ships with several of these, registered in
            // src/main/resources/META-INF/services
            throw new RuntimeException("Could not find any " + ClassLoaderHandler.class.getSimpleName()
                    + " subclasses registered on the classpath using the ServiceLoader mechanism");
        }
```

We're using the standard Maven Shade plugin to relocate the FastClasspathScanner to another package. We do this because Orbit is primarily designed to be a library that is used by others, and we don't like to introduce transitive dependencies to users. If they want to use the FastClasspathScanner in their own project, we don't want to introduce version clashes and such.
You can probably get an idea of what we're doing from the change here: https://github.com/orbit/orbit/commit/5188fc77bf2b5d3a145ef96e5375c3de3347838f#diff-de920684fd9adf848a0c051f7d7f1675

When the maven shade plugin relocates the classes, it does not handle the services correctly unfortunately. It's unfortunate that Maven doesn't do this, as I agree that is the ideal solution.

I see 2 options to resolve the issue purely in our project, but I don't think either make a whole lot of sense from a workflow point of view.

1. Write a custom maven plugin that relocates ```src/main/resources/META-INF/services/io.github.lukehutch.fastclasspathscanner.classpath.classloaderhandler.ClassLoaderHandler``` to  ```src/main/resources/META-INF/services/cloud.orbit.core.shaded.io.github.lukehutch.fastclasspathscanner.classpath.classloaderhandler.ClassLoaderHandler``` and rewrites the packages within the file itself. 
This would work, but writing a custom plugin is not really ideal.

2. Add our own ```src/main/resources/META-INF/services/cloud.orbit.core.shaded.io.github.lukehutch.fastclasspathscanner.classpath.classloaderhandler.ClassLoaderHandler``` to our project and maintain it. 
This is probably the easiest solution, but I worry about the long term maintenance of ensuring we remain in sync when we take new drops of FastClasspathScanner. If you relocated the loaders etc it would be very easy for us to miss, as it's not going to cause a direct compile failure or similar.

Given neither of these options seems ideal, I felt it may make sense to support a consuming project of FastClasspathScanner to be able to define custom ClassLoaders at runtime. It would allow us to solve our specific case where we'd actually just pass in the existing ClassLoaders you have, but it also potentially allows anyone to define a custom class loader and use it at runtime if they wanted some custom logic for class loading, rather than needing to essentially do it at build time with the META-INF/services feature.

I'm curious what your thoughts are.

Thanks,
Joe

**Comment by lukehutch** [comment]
Hi, thanks for explaining. I just pushed two changes to master:

1. URLClassLoaderHandler is now always enabled by default, whether or not ServiceLoader is working right.
2. I added the method registerClassLoaderHandler(ClassLoaderHandler extraClassLoaderHandler) to the FastClasspathScanner class, which you can use to manually register ClassLoaderHandlers before calling .scan().

Please test these and let me know if it works for you. If so, these will be in the next release.

However, based on your description, this is quite clearly a bug in Maven Shade, and you should report the bug there to get it fixed properly (then you won't even have to manually register anything). (ServiceLoaders are not widely used, so probably they just haven't come across this before.)  Shade really should rewrite any META-INF/services files it finds with applicable package naming changes. It should be a relatively straightforward fix.

**Comment by johnou** [comment]
@lukehutch I can confirm that with your latest changes ( 1.9.20-SNAPSHOT ) everything works as expected, thanks for picking this up so quickly! When do you plan to release 1.9.20?

**Comment by lukehutch** [comment]
You're welcome, thanks for verifying. I'll push out 1.9.20 right now then, since it's a minor change and I'll be traveling for most of the next couple of weeks.

Note that this change will only fix the problem for you in an automated way if the ClassLoader used by your project(s) is a subclass of URLClassLoader (which is the most common case), but it won't work with JBoss / WebLogic. I still recommend filing a bug with the Maven Shade plugin project.

**Comment by lukehutch** [comment]
Solved in 1.9.20.
